daffodil-core/src/main/scala/org/apache/daffodil/runtime2/generators/ParserGenerator.scala

```diff
@@ -0,0 +1,314 @@
+  }
+
+  def addComputations(child: ElementBase): Unit = {
+    val C = structs.top.C
+    val e = child.namedQName.local
+    val offsetComputation = s"    (char *)&${C}_compute_ERD_offsets.$e - (char *)&${C}_compute_ERD_offsets"
+    val erdComputation = s"    &${e}_ERD"
+    structs.top.offsetComputations += offsetComputation
+    structs.top.erdComputations += erdComputation
+  }
+
+  def addFieldDeclaration(context: ThrowsSDE, child: ElementBase): Unit = {
+    val definition = if (child.isSimpleType) {
+      import NodeInfo.PrimType
+      child.optPrimType.get match {
+        case PrimType.Long => "int64_t    "
+        case PrimType.Int => "int32_t    "
+        case x => context.SDE("Unsupported primitive type: " + x)
+      }
+    } else {
+      child.namedQName.local + "         "
+    }
+    structs.top.declarations += s"    $definition ${child.name};"
+  }
+
+  def viewCodeHeader: String = {
+    val structs = finalStructs.mkString("\n")
+    val header =
+      s"""#ifndef GENERATED_CODE_H
+         |#define GENERATED_CODE_H
+         |
+         |#include "common_runtime.h" // for InfosetBase
+         |#include <stdint.h>         // for int32_t
+         |
+         |// Return the root of an infoset to be used for parsing or unparsing
+         |
+         |extern InfosetBase *rootInfoset();
+         |
+         |// Define some infoset structures
+         |
+         |$structs
+         |#endif // GENERATED_CODE_H
+         |""".stripMargin
+    header
+  }
+
+  def viewCodeFile(rootElementName: String): String = {
+    val prototypes = this.prototypes.mkString("\n")
+    val erds = this.erds.mkString("\n")
+    val finalImplementation = this.finalImplementation.mkString("\n")
+    val code =
+      s"""#include "generated_code.h"
+         |#include <endian.h> // for be32toh, htobe32
+         |#include <errno.h>  // for errno
+         |#include <stddef.h> // for ptrdiff_t
+         |#include <stdio.h>  // for NULL, fread, fwrite, size_t, feof, ferror, FILE
+         |#include <string.h> // for strerror
+         |
+         |// Prototypes needed for compilation
+         |
+         |$prototypes
+         |
+         |// Metadata singletons
+         |
+         |$erds
+         |// Return the root of an infoset to be used for parsing or unparsing
+         |
+         |InfosetBase *
+         |rootInfoset()
+         |{
+         |    static $rootElementName    instance;
+         |    InfosetBase *root = &instance._base;
+         |    ${rootElementName}_ERD.initSelf(root);
+         |    return root;
+         |}
+         |
+         |// Methods to initialize, parse, and unparse infoset nodes
+         |
+         |static const char *
+         |eof_or_error_msg(FILE *stream)
+         |{
+         |    if (feof(stream))
+         |    {
+         |        static const char *error_msg = "Got EOF while expecting more input";
+         |        return error_msg;
+         |    }
+         |    else if (ferror(stream))
```

Comment:
ferror doesn't set errno. So the return strerror(errno) below ... is depending on something outside of this function having set errno appropriately. But fread/fwrite don't set errno.  Suggest we pass an arg to this specifying whether the error occurred in fread or fwrite, then we issue an error like "Error during data read for %s" where %s comes from the ERD (which we should also pass.)